### DISCUSSION TITLE
Email alert have option to skip sending when dashboard/chart was unable to load successfully

### DISCUSSION THREAD

**Discussion by unnyns-307** [apache/superset#22312] [Ideas]
Hello all, after we have implemented to use Superset in our organization. We also apply Superset to send email report to many business teams and it would be great if we can have option to prevent email report sending when the dashboard/chart cannot be load successfully.

and we got a lot of feedback that this option will give better user experience for business user while we still can investigate the root cause of alert unable to send in superset UI at alert's log page without showing error on email report to end users.

### Suggested option: 
have an **option in Alerts and Reports setting** for example, 
- [x] Don't send report when there is no result
- [x] Only send report when content load successfully

![image](https://user-images.githubusercontent.com/82924994/205260076-9e5e1f8f-de01-417f-a2a4-fc077a6e4d51.png)

**Comment by rusackas** [comment]
Pinging @yousoph for input/consideration here. This seems pretty reasonable to me, if there's a good place to put a try/catch in the logic. 

I guess one question here (for anyone's input) is: if the content doesn't load correctly, and the report fails to send, should there be some secondary notification sent to the alert/report author? Probably better to somehow be notified of the failure than to get a "where's my report?" question from a coworker/customer/boss/etc.

**Reply by eschutho** [reply]
Thanks for the suggestions @unnyns-307. For the two options that you’re proposing, just for a bit of clarification:
1) don’t send a report if there is no result 
Does this mean that the dashboard has no charts? Or didn’t load properly?
2) only send when content loads properly
Similar question to the above. Should this send a report if one out of say 10 charts didn’t load properly? Like all or nothing?


We have been thinking of something similar.

**Reply by unnyns-307** [reply]
Hello @eschutho, thank you for looking into this. 
Actually, for both options refer to the behavior cannot load since there is some error on chart(s) of the dashboard.
If 'there is no result' mean the chart load successfully but it has no related data to show, then it's okay send the alert as normal.

I have seen some other visualization tool have additional option in their alert feature to prevent sending the alert when there is no related data show on chart as well. I think it's not a significant to have in the first phase, but it surely is good for user experience if we could provide this ability too.

Hence, we would like to suggest that the option should prevent alert sending when there is at least one of chart in dashboard cannot be loaded properly.  It would be better if we could have another email sending the error to the alert's owner as @rusackas suggested too.

**Reply by rusackas** [reply]
CC @yousoph for context.

**Reply by yousoph** [reply]
@unnyns-307 thanks for the suggestion! Makes sense for charts to not send the email if the chart doesn't load successfully... but for dashboards, we were thinking about sending the report of the partially loaded dashboard if some of the charts have loaded but others haven't. Curious to hear what your thoughts (or anybody else's!) are here. cc @Rohit-pawar902 

To @rusackas 's question, I do think it makes sense to alert the owner of the report if the chart doesn't load and the report doesn't send.

**Reply by unnyns-307** [reply]
@yousoph Thanks for sharing your thought. I would say end-user would mainly focus on the result if there is one chart on dashboard with error, user would still question is there any problem or incident with the system/ is the data updated properly. 

We think it'd be nice if the alert could check that all chart in the dashboard is loaded successfully and capture the result from that latest dashboard's refresh. 

- I notice that currently Superset have a minor limitation to send alert with capture of most updated data if we didn't set the time interval to refresh at dashboard level or at dataset level before the time alert should be sent. Let's say for now Superset sends alert with result from the last time dashboard is refreshed which sometimes it's not the last updated data because the dashboard hasn't been refreshed for a while. 

So I found somehow, it's harder to manage to send the most recent result when there is multiple alerts sending with different time schedule. The workaround might be we need to set database to refresh at the smallest time unit but I'm not sure it would be drawback for the performance of application or not.

However, loading all chart again every time the schedule run could impact the performance of application and workload at the backend also. Because if the dashboard take a long time to load, the schedule could be delayed too but I believe if we could find solution for this problem it could help the use case from @Rohit-pawar902 provided below as well https://github.com/apache/superset/discussions/22312#discussioncomment-4324977

Please feel free to discuss what could be suitable alternative way for our case

**Comment by rohitpawar2811** [comment]
I think email avoid feature should be their But I am pointing my case when there is some sort of error on chart or something else like that.
Actually in my case I have big query which takes time longer than usually that's some time on email-alert It would not load properly on dashboard .

![Screenshot from 2022-12-06 22-01-24](https://user-images.githubusercontent.com/72196393/205970684-11ae8133-09c8-42f4-9752-1cb69f1aa33a.png)
